### PR TITLE
Update Method Invocations To Be JRE 8 Compatible

### PR DESCRIPTION
**Pull request by isomarcte**
More recent JREs have added new methods, or overloads of existing methods, which create runtime issues when compiling Scala with a newer JRE and then running with an older JRE, namely JRE 8.

This commit addresses,

* `java.lang.StringBuilder#isEmpty` 
    * `isEmpty` got added as a default method to `CharSequence`. In the file `Scanners.scala` an implicit class in the companion object also provided `isEmpty`, but when compiling with a newer JRE the scala compiler will link to the method on the class. Since this method doesn't exist in older JREs, this creates a runtime failure.
* `java.nio.Buffer` related methods 
    * The `Buffer` superclass declares many of the core methods that the various subtypes implement. In the superclass, these methods have a fluent interface, each returning the class itself. However, on `Buffer` they are defined to return the type `Buffer`, as opposed to the concrete type, e.g. `ByteBuffer`. More recent JREs overrode the implementation of many of these methods in the concrete classes to return the concrete type, e.g. `(bb: ByteBuffer).flip(): ByteBuffer`. This is much nicer, but when compiling the compiler with newer JREs the scala compiler will link against the overloaded method with the more specific return type. This causes a runtime failure when running the scala compiler or related tools on an older JRE. 
    * This commit replaces the invocations of these methods with wrapper methods which emulate the newer behavior and avoid the runtime failures. 
    * These methods include, `flip`, `clear`, `position`, and `limit`.

I'm honestly not sure how to write tests for this in a comprehensive manner (suggestions welcome!). I recently became the maintainer of the scala package for ArchLinux and I encountered the issue when building that package with a newer JRE and then running it with JRE 8. This is a simple, but imprecise, way to test this.

Here is an example stack trace from a failure.

```
java.lang.NoSuchMethodError: java.nio.ByteBuffer.position(I)Ljava/nio/ByteBuffer;
        at scala.tools.nsc.symtab.classfile.ReusableDataReader.skip(ReusableDataReader.scala:125)
        at scala.tools.nsc.symtab.classfile.ClassfileParser$ConstantPool.<init>(ClassfileParser.scala:224)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.$anonfun$parse$2(ClassfileParser.scala:174)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.$anonfun$parse$2$adapted(ClassfileParser.scala:160)
        at scala.reflect.internal.util.ReusableInstance.using(ReusableInstance.scala:39)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.$anonfun$parse$1(ClassfileParser.scala:160)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.pushBusy(ClassfileParser.scala:143)
        at scala.tools.nsc.symtab.classfile.ClassfileParser.parse(ClassfileParser.scala:159)
        at scala.tools.nsc.symtab.SymbolLoaders$ClassfileLoader.doComplete(SymbolLoaders.scala:342)
        at scala.tools.nsc.symtab.SymbolLoaders$SymbolLoader.$anonfun$complete$2(SymbolLoaders.scala:249)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.reflect.internal.SymbolTable.informingProgress(SymbolTable.scala:85)
        at scala.tools.nsc.symtab.SymbolLoaders$SymbolLoader.complete(SymbolLoaders.scala:247)
        at scala.reflect.internal.Symbols$Symbol.completeInfo(Symbols.scala:1563)
        at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1535)
        at scala.reflect.internal.Symbols$Symbol.initialize(Symbols.scala:1724)
        at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1650)
        at scala.tools.nsc.Global$Run.<init>(Global.scala:1236)
        at scala.tools.nsc.interpreter.IMain.liftedTree1$1(IMain.scala:152)
        at scala.tools.nsc.interpreter.IMain.global$lzycompute(IMain.scala:151)
        at scala.tools.nsc.interpreter.IMain.global(IMain.scala:142)
        at scala.tools.nsc.interpreter.IMain.initializeCompiler(IMain.scala:140)
        at scala.tools.nsc.interpreter.shell.ILoop.$anonfun$run$2(ILoop.scala:955)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.interpreter.IMain$.withSuppressedSettings(IMain.scala:1420)
        at scala.tools.nsc.interpreter.IMain.withSuppressedSettings(IMain.scala:106)
        at scala.tools.nsc.interpreter.shell.ILoop.$anonfun$run$1(ILoop.scala:954)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.interpreter.shell.ReplReporterImpl.withoutPrintingResults(Reporter.scala:64)
        at scala.tools.nsc.interpreter.shell.ILoop.run(ILoop.scala:954)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:87)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Exception in thread "main" java.lang.NullPointerException
        at scala.tools.nsc.CompilationUnits$CompilationUnit.<init>(CompilationUnits.scala:44)
        at scala.tools.nsc.CompilationUnits$CompilationUnit.<init>(CompilationUnits.scala:45)
        at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.compile(IMain.scala:735)
        at scala.tools.nsc.interpreter.IMain.bind(IMain.scala:554)
        at scala.tools.nsc.interpreter.IMain.bind(IMain.scala:590)
        at scala.tools.nsc.interpreter.IMain.$anonfun$quietBind$1(IMain.scala:589)
        at scala.tools.nsc.interpreter.shell.ReplReporterImpl.withoutPrintingResults(Reporter.scala:64)
        at scala.tools.nsc.interpreter.IMain.quietBind(IMain.scala:589)
        at scala.tools.nsc.interpreter.shell.ILoop.interpretPreamble(ILoop.scala:922)
        at scala.tools.nsc.interpreter.shell.ILoop.$anonfun$run$3(ILoop.scala:963)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.interpreter.shell.ILoop.echoOff(ILoop.scala:90)
        at scala.tools.nsc.interpreter.shell.ILoop.$anonfun$run$2(ILoop.scala:963)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.interpreter.IMain$.withSuppressedSettings(IMain.scala:1420)
        at scala.tools.nsc.interpreter.IMain.withSuppressedSettings(IMain.scala:106)
        at scala.tools.nsc.interpreter.shell.ILoop.$anonfun$run$1(ILoop.scala:954)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.tools.nsc.interpreter.shell.ReplReporterImpl.withoutPrintingResults(Reporter.scala:64)
        at scala.tools.nsc.interpreter.shell.ILoop.run(ILoop.scala:954)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:87)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

### CODE REVIEW

**Comment by som-snytt** [comment]
This is why releases are built under JDK8 or equivalently `--release:8`.

It must be dubious (a "fool's errand") to support compiling against later JDK and run under earlier.

Note that we don't close contributions tout court, unlike https://github.com/lampepfl/dotty/pull/17249.

**Comment by isomarcte** [comment]
@som-snytt Yeah, I agree it should always be compiled with JDK 8. That is a change I made in the package and intend to keep regardless of this PR's outcome. I'm happy to close if the changes aren't wanted.

Out of curiosity, is there a reason that scalac doesn't use the `-release:8` flag when compiling itself by default in the `bulid.sbt`?

**Comment by som-snytt** [comment]
I'm sure @SethTisue will chime in with an expert opinion.

They build against newer JDK, perhaps to reduce tech debt. `--release:8` would make that less effective.

I just tweaked `partest` to cope a bit better with an impedance mismatch between the JDK in use and JDK it's compiling for.

It would be nice if the build output were multi-versioned: I'm not sure there are opportunities for source differences, but there certainly could be fruitful differences in class output per version.

While making a casual contribution to the project, I often accidentally build under one JDK and test under another. (Usually because I was testing something else and had set my "ambient" JDK.)

**Comment by SethTisue** [comment]
> I'm sure @SethTisue will chime in with an expert opinion

(maybe, but probably not before 2.13.11 ships. need to think on this)

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala [hunk 1]
```diff
@@ -1345,7 +1350,12 @@ trait Scanners extends ScannersCommon {
 
     // disallow trailing numeric separator char
     def checkNoTrailingSeparator(): Unit =
-      if (!cbuf.isEmpty && isNumberSeparator(cbuf.last))
+      // JDK 15 added this to CharSequence, which java.lang.StringBuilder
+      // implements. This means we need to take care to invoke only our
+      // version so that we don't emit bytecode which will try to link against
+      // java.nio.StringBuilder#isEmpty which would fail at runtime when
+      // running on a JRE < 15.
+      if (!(new Cbuf.StringBuilderOps(cbuf).isEmpty) && isNumberSeparator(cbuf.last))
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala:1358]: As compiled, the call is not allocating:
```
scala> :javap scala.tools.nsc.ast.parser.Scanners$Scanner#checkNoTrailingSeparator
        11: invokevirtual #1091               // Method scala/tools/nsc/ast/parse/Cbuf$StringBuilderOps$.isEmpty$extension:(Ljava/lang/StringBuilder;)Z
```

**Comment by som-snytt** [comment]
We don't want to address evolving platform API with code tweaks because building against new platforms is useful. (Cf https://github.com/scala/scala/pull/10222 and related tickets.) (You'd have to revert tweaks if support for JDK 8 was dropped.)

Besides `--release:8`, it would be useful to know when extensions are no longer used.

The paradigmatic clash was `string.lines`, which became `linesIterator`. Some people would prefer to compile against a fixed `ct.sym`.

**Pull request closed without merging**